src/main/java/org/jitsi/jigasi/SipCallTransformerMonitor.java

```diff
@@ -0,0 +1,228 @@
+     *
+     * @param pkt the packet.
+     * @return the original packet.
+     */
+    @Override
+    public RawPacket transform(RawPacket pkt)
+    {
+        lastOutgoingActivity = System.currentTimeMillis();
+        seenSSRCs.add(pkt.getSSRCAsLong());
+
+        return pkt;
+    }
+
+    @Override
+    public RawPacket reverseTransform(RawPacket pkt)
+    {
+        seenSSRCs.add(pkt.getSSRCAsLong());
+
+        return pkt;
+    }
+
+    /**
+     * Implements {@link TransformEngine#getRTPTransformer()}.
+     */
+    @Override
+    public PacketTransformer getRTPTransformer()
+    {
+        return this;
+    }
+
+    /**
+     * Implements {@link TransformEngine#getRTCPTransformer()}.
+     * <p>
+     * This <tt>TransformEngine</tt> does not transform RTCP packets.
+     */
+    @Override
+    public PacketTransformer getRTCPTransformer()
+    {
+        return rtcpTransformer;
+    }
+
+    /**
+     * Checks and filters RTCP.BYE.
+     */
+    public class RTCPTransformer
+        extends SinglePacketTransformerAdapter
+    {
+        RTCPTransformer()
+        {
+            super(RTCPPacketPredicate.INSTANCE);
+        }
+
+        @Override
+        public RawPacket transform(RawPacket pkt)
+        {
+            RTCPIterator it = new RTCPIterator(pkt);
+            while (it.hasNext())
+            {
+                ByteArrayBuffer baf = it.next();
+                int type = RTCPUtils.getPacketType(baf);
+                long ssrc = RawPacket.getRTCPSSRC(baf);
+
+                // Filter RTCP.BYE for streams we don't know about
+                if (!seenSSRCs.contains(ssrc) && type == 203)
+                {
+                    it.remove();
+                }
+            }
+
+            if (pkt.getLength() > 0)
+            {
+                lastOutgoingActivity = System.currentTimeMillis();
+                return pkt;
+            }
+
+            return null;
+        }
+    }
+
+    /**
+     * Called periodically to check when was the last outgoing RTP or RTCP,
+     * if it was longer than a threshold we send hole punch packets to keep
+     * media, as some providers can stop media till they see something coming.
+     * In this case all web participants can be muted and the bridge will not be
+     * forwarding any media so we can stop sending any media to the sip side
+     * and the incoming media can stop and participants will stop hearing that
+     * side.
+     */
+    private class KeepAliveIncomingMedia
+        extends PeriodicRunnable
+    {
+        public KeepAliveIncomingMedia(long period)
+        {
+            super(period);
+        }
+
+        @Override
+        public void run()
+        {
+            super.run();
+
+            try
+            {
+                if (System.currentTimeMillis()
+                        - lastOutgoingActivity > getPeriod())
+                {
+                    handler.sendHolePunchPacket(stream, MediaType.AUDIO);
```

Comment:
I see a few problems or potential problems here.
1. Hole-punch packets are meant for the start of a session. Looking at the jitsi code that sends them, they will have a valid PT and SSRC, but they have a hardcoded sequence number (0x29e). Sending such a packet in the middle of a stream could lead to problems (e.g. it could brake the SRTP session). 

2. We're sending a single packet, and it could be lost.

3. The timeout period is the same as period of the executable -- 25 seconds. This means that in some cases we'll start sending keep alive packets up to 50 seconds after the last activity. IIRC the threshold we're after is 35 seconds.

Is there a standarsized or widely-used approach for sending keep alive packets with RTP? [This document](https://tools.ietf.org/html/rfc6263#section-5) recommends using RTCP, e.g. we could send empty RTCP RRs (or something else?) instead.

Reply:
Talked offline and will try implementing this with different ssrc and RTP and PT 13(comfort noise).